File: Graphs/BFS/399. Evaluate Division/solution.py
# a/b = a/aa * aa/b -> think this as a graph from a -> aa -> b
# use BFS to explore answer
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        results = defaultdict(lambda: -1)
        graph = defaultdict(list)
        for i, (a, b) in enumerate(equations):
            results[(a,a)] = 1
            results[(b,b)] = 1
            results[(a,b)] = values[i]
            results[(b,a)] = 1 / values[i]

            graph[a].append(b)
            graph[b].append(a)
            
        def BFS(x, y):
            queue = deque([(x,y,1)])
            visited = set()
            while queue:
                for _ in range(len(queue)):
                    a, b, curr = queue.popleft()

                    if (a, b) in visited: continue
                    visited.add((a,b))

                    if (a,b) in results:
                        return results[(a,b)] * curr

                    for aa in graph[a]:
                        # a/b = a/aa * aa/b
                        if results[(a,aa)] == -1: continue
                        queue.append([aa,b,curr*results[(a,aa)]])
            return -1

        res = []
        for x, y in queries:
            if (x not in graph) or (y not in graph):
                res.append(-1)
            elif x == y:
                res.append(1)
            else:
                v = BFS(x, y)
                results[(x,y)] = v
                results[(y,x)] = 1/v
                res.append(v)
        return res

File: Graphs/BFS/399. Evaluate Division/solution.cpp
// DFS solution
class Solution {
public:
    double dfs(string a, string b, unordered_map<string, vector<string>>& graph, 
               map<pair<string,string>, double>& quotient, unordered_set<string>& visited) {
        if (a == b) return 1.0;
        
        visited.insert(a);
        
        for (auto nxt : graph[a]) {
            if (visited.count(nxt)) continue;
            
            double x = dfs(nxt, b, graph, quotient, visited);
            if (x > 0) {
                return x * quotient[{a, nxt}];
            }
        }
        
        return -1.0;
    }
    
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        unordered_map<string, vector<string>> graph;
        map<pair<string,string>, double> quotient;
        
        for (int i = 0; i < equations.size(); i++) {
            string a = equations[i][0], b = equations[i][1];
            graph[a].push_back(b);
            graph[b].push_back(a);
            
            quotient[{a, b}] = values[i];     // a/b = value
            quotient[{b, a}] = 1.0 / values[i]; // b/a = 1/value
        }
        
        vector<double> res;
        for (auto query : queries) {
            string a = query[0], b = query[1];
            
            // Check if both variables exist in the graph
            if (graph.find(a) == graph.end() || graph.find(b) == graph.end()) {
                res.push_back(-1.0);
                continue;
            }
            
            if (a == b) {
                res.push_back(1.0);
            } else {
                unordered_set<string> visited;
                double x = dfs(a, b, graph, quotient, visited);
                res.push_back(x);
            }
        }
        return res;
    }
};